src/Analyzer/Passes/FuseFunctionsPass.cpp

```diff
@@ -0,0 +1,156 @@
+#include <Analyzer/Passes/FuseFunctionsPass.h>
+
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <Functions/FunctionFactory.h>
+
+#include <AggregateFunctions/IAggregateFunction.h>
+
+#include <Analyzer/InDepthQueryTreeVisitor.h>
+#include <Analyzer/FunctionNode.h>
+#include <Analyzer/ConstantNode.h>
+
+#include <DataTypes/DataTypesNumber.h>
+#include <DataTypes/DataTypeTuple.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+namespace
+{
+
+class FuseFunctionsVisitor : public InDepthQueryTreeVisitor<FuseFunctionsVisitor>
+{
+public:
+
+    static bool matchFunctionName(const String & name)
+    {
+        return name == "count" || name == "sum" || name == "avg";
+    }
+
+    void visitImpl(QueryTreeNodePtr & node)
+    {
+        auto * function_node = node->as<FunctionNode>();
+        if (!function_node || !function_node->isAggregateFunction() || !matchFunctionName(function_node->getFunctionName()))
+            return;
+
+        const auto & arguments = function_node->getArgumentsNode()->getChildren();
```

Comment:
This `getChildren` method should not be used anywhere except visitors. You can use `getArguments` method that will return reference to `ListNode`.

Reply:
Changed to `getArguments().getNodes()`